runtime/src/prioritization_fee.rs

```diff
@@ -0,0 +1,354 @@
+
+    // Default to `false`, set to `true` when a block is completed, therefore the minimum fees recorded
+    // are finalized, and can be made available for use (e.g., RPC query)
+    is_finalized: bool,
+
+    // slot prioritization fee metrics
+    metrics: PrioritizationFeeMetrics,
+}
+
+impl Default for PrioritizationFee {
+    fn default() -> Self {
+        PrioritizationFee {
+            min_transaction_fee: u64::MAX,
+            min_writable_account_fees: HashMap::new(),
+            is_finalized: false,
+            metrics: PrioritizationFeeMetrics::default(),
+        }
+    }
+}
+
+impl PrioritizationFee {
+    /// Use `sanitized_tx` to update self for minimum transaction fee in the block and minimum fee for each writable account.
+    pub fn update(
+        &mut self,
+        sanitized_tx: &SanitizedTransaction,
+    ) -> Result<(), PrioritizationFeeError> {
+        let (_, update_time) = measure!(
+            {
+                let account_locks = sanitized_tx
+                    .get_account_locks()
+                    .or(Err(PrioritizationFeeError::FailGetTransactionAccountLocks))?;
+
+                let priority_details = sanitized_tx
+                    .get_transaction_priority_details()
+                    .ok_or(PrioritizationFeeError::FailGetTransactionPriorityDetails)?;
+
+                if priority_details.priority < self.min_transaction_fee {
+                    self.min_transaction_fee = priority_details.priority;
+                }
+                for write_account in account_locks.writable {
+                    self.min_writable_account_fees
+                        .entry(*write_account)
+                        .and_modify(|write_lock_fee| {
+                            *write_lock_fee =
+                                std::cmp::min(*write_lock_fee, priority_details.priority)
+                        })
+                        .or_insert(priority_details.priority);
+                }
+
+                self.metrics
+                    .increment_total_prioritization_fee(priority_details.priority);
+            },
+            "update_time",
+        );
+
+        self.metrics
+            .increment_total_update_elapsed_us(update_time.as_us());
+        Ok(())
+    }
+
+    /// Accounts that have minimum fees lesser or equal to the minimum fee in the block are redundant, they are
+    /// removed to reduce memory footprint.
+    pub fn prune_irrelevant_writable_accounts(&mut self) {
+        self.metrics.total_writable_accounts_count = self.get_writable_accounts_count() as u64;
+        self.min_writable_account_fees
+            .retain(|_, account_fee| account_fee > &mut self.min_transaction_fee);
+        self.metrics.relevant_writable_accounts_count = self.get_writable_accounts_count() as u64;
+    }
+
+    pub fn mark_block_completed(&mut self) -> Result<(), PrioritizationFeeError> {
+        if self.is_finalized {
+            return Err(PrioritizationFeeError::BlockIsAlreadyFinalized);
```

Comment:
This seems like more of an assertion. Is this something expected to happen, ever?

Reply:
above modification makes this code path clearer ig